Lib/dataclasses.py

```diff
@@ -1202,7 +1202,14 @@ def _get_slots(cls):
         # A class which does not define __slots__ at all is equivalent
         # to a class defining __slots__ = ('__dict__', '__weakref__')
         case None:
-            yield from ('__dict__', '__weakref__')
+            # Except for special cases, inheriting from them do not set
+            # any slots at all:
+            slots = ['__dict__', '__weakref__']
+            if getattr(cls, '__weakrefoffset__', -1) == 0:
+                slots.remove('__weakref__')
+            if getattr(cls, '__dictrefoffset__', -1) == 0:
+                slots.remove('__dict__')
+            yield from slots
         case str(slot):
             yield slot
```

Comment:
AFAIK it is possible for a C extension type to have a `__slots__` attribute. It doesn't usually make sense to do so, since it isn't necessary for creating slots on an extension type, but someone might do it in order to make their type look more like a Python-defined one? In such a case I think the code here will still give the wrong answer; perhaps we should use the `__weakrefoffset__` and `__dictoffset__` detection code in all cases, not only in the "no `__slots__`" case?

Reply:
>  It doesn't usually make sense to do so, since it isn't necessary for creating slots on an extension type, but someone might do it in order to make their type look more like a Python-defined one?

I am not sure that this is even a correct thing to do, honostely. I propose to open a new issue about this case and discuss this separately. First, because it is not related to the original issue. Second, because it is rather rare. Third, because we have a feature freeze coming and I want to have this merged and released before the first beta :)